DM-26658: Re-implement the Formatter class

timj opened the pull request:
* Use entirely new class: FormatterV2
* Provide FormatterV1inV2 shim class.
* Since FormatterV2 needs access to the cache, moved all the cache code to FormatterV2 itself out of Datastore.
* Three read methods can be provided, two write methods.
* Can now read content from within zip archives.

For read:

* `read_from_uri` gives you the full URI to do with whatever you want but *might* give you a locally file if it was in the cache.
* `read_from_stream` gives you a file handle for the resource.
* `read_from_local_file` gives you a guaranteed local file.

Flags declare which of those 3 are available for a given formatter to allow the `FormatterV2.read` method to pick a suitable option based on what it knows (or if a zip file is in play, in which case `read_from_uri` can't be called).

For write:

* `to_bytes` is called first (can be Not Implemented)
* Falls back to `write_local_file`

The default implementation of `write_local_file` calls `to_bytes`.

All daf_butler formatters ported to V2. There are some V1 test formatters since otherwise nothing will test `FormatterV1inV2`.

This mostly seems to work but there are some things I still need to do:

- [x] There is some code duplication within FormatterV2 that I should tidy up.
- [x] I agreed to add any notes from sub exceptions when I re-raise so that notes are more visible in the final stack trace.
- [x] Work out how to prevent storage class conversion before handing off to a formatter.


## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Write new Formatter implementation

Entirely new FormatterV2 with a wrapper class to support
legacy Formatter implementations. No longer subclass the
read and write method but instead subclass

* read_from_uri
* read_from_stream
* read_from_local_file
* to_bytes
* write_local_file

timj committed:
Reset storage class singleton between tests

This mostly wasn't a problem but there is a test somewhere that
is modifying StructuredComposite such that if it runs before
the transfer tests the transfer test fails because the converter
definitions have been stripped.

timj committed:
Change approach to kwargs merging

Using Config is fine when merging dicts but it breaks if
a top-level kwarg is a Mapping but should not be merged.
This was being triggered when formatter constructors
switched to using dataId=DataCoordinate. Reorganize
the merging such that only keys that appear in both
kwargs are merged using Config.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +60,927 @@
+class FormatterV2(metaclass=ABCMeta):
+    """Interface for reading and writing datasets using URIs.
+
+    The formatters are associated with a particular `StorageClass`.
```
One or more storage classes, right?  `ParquetFormatter` speaks can be used with multiple storage classes.

timj replied:
A storage class is associated with one formatter in configuration.  I need to think about how Formatter can say what it supports. The current situation is:

* On read the formatter can see what the requested storage class is and generate that directly.
* On write the datastore does a type conversion before sending the in-memory dataset to the formatter

We would need something like a `FormatterV2.can_accept(in_memory_dataset)` API that would allow a formatter to say it's all under control, bypass the type conversion. Maybe it has to take the storageClass instead. Need to think. It's not solely a case of "can you accept X" it's also an assertion that the formatter can read it back as the registry storage class python type.

timj replied:
I have implemented `can_accept` so that @erykoff can prevent conversions to DataFrame on write.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +60,927 @@
+
+
+class FormatterV2(metaclass=ABCMeta):
+    """Interface for reading and writing datasets using URIs.
```
One thing I was a little uncomfortable with in the original `Formatter` interface is that the same class is responsible for both reading and writing, but the state the class wants to hold for reading is not generally the same as the state it wants to hold for writing.  It's nice to have one class name to configure rather than two, but other than that I think reading and writing are best kept separate.

So, while I realize this would be another pretty big change (apologies for not raising the possibility earlier), what do you think about making the `write` interface go entirely through `classmethod`s, so that constructing an instance is something you do only for reads?  The reason I'd let the instance state be for reads is to give the formatter the URL and `DatasetRef` at construction to let it do some work up front, and then support multiple calls to single-component or dict-of-parameters reads that it may be able to optimize internally (e.g. using [fsspec's buffering/caching file-like object wrappers](https://filesystem-spec.readthedocs.io/en/latest/api.html#readbuffering)).  Getting multiple components or sets of parameters would require new interfaces higher up, too, of course, but I think that's easier than the formatter side.

If we don't go this route, I think we want to be very clear about these being two-state instances that are constructed in different ways for read and write and may have different attributes set as a result.  And we really want to limit to just two states.

timj replied:
Writing currently needs to be an instance method because the write parameters are read from configuration and attached to the formatter instance by the formatter factory.

I've only just read the second half of your second paragraph where you are suggesting that `formatter.read()` might want to be called multiple times with different components. Datastore isn't set up to be able to do that at the moment because butler only asks for one component for one dataset or else all components from a disassembled dataset.

TallJimbo replied:
I do think reading N components might be useful, but I don't have a concrete use case in mind and I could imagine it being better implemented by being able to pass a list of components rather than by reworking what is considered formatter instance state.  It's not something I'd go add at a high level right now, just something to keep in mind as a possibility (e.g. is the formatter interface such that we could add a method for that with a default implementation and override it in just some concrete formatters, rather than having to change them all again).

timj replied:
Now that the user subclasses something like `read_from_uri(uri, component)` and doesn't subclass `read` itself then we could pass multiple components to `read` and have that by default be N calls to `read_from_uri`.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +60,927 @@
+        """
+        raise NotImplementedError("This formatter does not know how to read a local file.")
+
+    def read_from_bytes(self, serialized_bytes: bytes, component: str | None = None) -> Any:
```
I think a "read from file-like object" is more general than "read from bytes" and a complete replacement for it.  I wonder if it'd also be a replacement either of the other methods or something they could delegate to by default, but I haven't wrapped my head around the caching enough to have a sense for whether that's the case.

timj commented:
For the record, we have been asked to support reading a file from within a zip file (where a single zip file is registered as multiple datasets). This might affect some of our thinking here as to the best interface.

timj committed:
Catch warnings from TAI to UTC conversion for future dates

timj commented:
The `.name` property used for stream handles requires lsst/resources#89.

timj committed:
Attach notes to re-raised exception if the formatter attached any

timj committed:
Add news fragment

timj committed:
Update the documentation for FormatterV2

timj committed:
Remove the old Formatter V1 implementations

timj committed:
Allow a formatter to declare it can accept a python type without coercion

timj commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1058 @@
+        # local file implementation which will trigger a cache write.
+        if self.can_read_from_uri:
+            with contextlib.suppress(FormatterNotImplementedError):
+                return self.read_directly_from_possibly_cached_uri(
```
I wonder if we should trap explicitly the no component / no parameters case and not call the read_from_uri here but defer it to the later situation where we have downloaded and possibly cached the file and then can pass the local URI to read_from_uri. That might be better than having the formatter author decide for themselves whether they might want to use the cached file instead.

TallJimbo replied:
If we do run into a case where we find it's important to download and cache the full file in order to do a component read (and it's not just because the formatter doesn't implement partial reads for that component), I think the formatter won't really able to know that any better than the datastore.  And I don't think those cases are likely to begin with, so simplicity arguments probably should carry the day.

In the scenarios where I imagine both a component of a dataset and the full dataset being inputs to a single `Quantum`, we actually don't want to bring down and cache the full dataset if the component is requested first (at least not automatically), because there's good chance the component is only being requested because looking at that might be enough for the task to raise `NoWorkFound`, so the full dataset might never be read.

timj replied:
I was thinking of the non-component read case since that one will need the full file anyhow (if there are no parameters). If a formatter author believes that it's always better to read from the local file then they should only implement `read_from_local_file`. What we don't want is for an author to implement `read_from_uri` by transferring the whole file locally and then opening it -- that would break caching. I think I need to add that to the documentation.

As currently implemented, if someone asks for a component and `read_from_uri` can handle a component read without downloading the whole file, then the whole file won't be cached.

timj replied:
I have changed the code so that read_from_uri can be called later after downloading to the cache if `read_from_uri` raised not implemented error. What do you think of always downloading the file to local storage (as we do now) if there is no component and not parameters specified?

timj replied:
Modified to do the download and cache if there are no components or parameters specified.

timj commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1058 @@
+        raise FormatterNotImplementedError("This formatter does not know how to read a file.")
+
+    def read_from_stream(
+        self, stream: BinaryIO | ResourceHandleProtocol, component: str | None = None
```
Is it worth passing in the expected size to this method (and read_from_uri) in case they want to decide that it would be more efficient to read from a local file and raise in here? (which would need to be caught above).

TallJimbo replied:
I think so, not because I think it's likely to be used, but because adding a new argument in the future would be a breaking change and it seems pretty harmless now.

timj replied:
Added to `read_from_uri` as well. Not added to `read_from_local_file` on the basis that there isn't really much choice even if it's a large file and also if they really need to know the actual size they can ask easily enough and performance shouldn't be an issue. Maybe I should add it for symmetry though.

timj replied:
I also added it to `read_from_local_file`...

timj committed:
Test stringification of formatter

The tests still use V1 though

timj committed:
Update only resources for docker container

Else tests do not pass.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        The dataset associated with this formatter. Should not be a component
+        dataset ref.
+    write_parameters : `dict`, optional
+        Any parameters to be hard-coded into this instance to control how
```
"hard-coded" seems like a strange choice of words for something that seems explicitly configurable.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        Any parameters to be hard-coded into this instance to control how
+        the dataset is serialized.
+    write_recipes : `dict`, optional
+        Detailed write Recipes indexed by recipe name.
```
Is Recipes capitalization intentional?  Is there a type this was supposed to link to?

timj replied:
That was not intentional.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+            cache_manager = DatastoreDisabledCacheManager(None, None)
+        return cache_manager
+
+    def read(
```
Might want to use `@typing.final` to enforce what the docs say.

timj replied:
Great idea. Except `TypelessFormatter` only works because it overrides `read`...

TallJimbo replied:
I think that reinforces my thinking that we should move away from having a JSON formatter and a single YAML formatter that try to do everything, and specialize for Pydantic/dataclasses,etc.  But not on this ticket.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        """
+        raise FormatterNotImplementedError("This formatter does not know how to read a local file.")
+
+    def write(
```
Might want to use `@typing.final` to enforce what the docs say.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+    (since it allows the possibility for a subset of the data file to be
+    accessed when parameters or components are specified) but it will not
+    update the local cache. If the entire contents of the remote file are
+    being accessed consider raising `FormatterNotImplementedError` and falling
```
Did you consider having it _return_ `NotImplemented` instead, like e.g. `__add__` and `__radd__` can?

timj replied:
I did not think about that.  That's presumably more efficient than a try/except and we'd have `read` raise `FormatterNotImplementedError` if nothing worked.

timj replied:
I tried this and it does work although:

* End up with a lot of extra `if result is not NotImplemented` checks
* It's not an `object` so changing the return type from `Any` to `object` means you have to change all the `read_from_*` methods to be explicitly returning `types.NotImplementedType | object` (or we can make a type alias).

What do you think? returning but then complicating the return type? Or returning and `Any`? or keeping it with `raise`?

timj replied:
Now return `NotImplemented` but switched back to `Any` in the type annotation.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+    method but instead to implement `write_local_file` and, optionally,
+    `to_bytes`. For local URIs the system will always call `write_local_file`
+    to ensure atomic writes are implemented. For remote URIs with local caching
+    disabled, `to_bytes` will be called first.
```
Implication here is that for remote URIs with local caching enabled, `write_local_file` will also be called first.  Is that right?

timj replied:
Yes. The code calls write_local_file if the file is local or the ref should be cached. The default implementation of write_local_file is to call to_bytes.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+    unsupported_parameters: ClassVar[Set[str] | None] = frozenset()
+    """Set of read parameters not understood by this `Formatter`. An empty set
+    means all parameters are supported.  `None` indicates that no parameters
+    are supported. These param (`frozenset`).
```
Unfinished last sentence.

TallJimbo replied:
Might also be clearer to explicitly state that these are read parameters _declared by a storage class that is otherwise supported by this formatter_.  Even though I should know better it took me a moment to figure out what set of parameters this represented a difference from.

timj replied:
There is an underlying assumption that if there are multiple storage classes that are configured to use a specific formatter that they have compatible parameter definitions. I probably should also add that a storage class delegate would be required to handle parameters that the formatter can not support.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+    how a dataset is serialized. `None` indicates that no parameters are
+    supported."""
+
+    default_extension: str | None = None
```
Not a `ClassVar`, then?

timj replied:
It should be. I forgot (mostly because it wasn't defined that way originally for Formatter)

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+    """Declare whether `read_from_stream` is available to this formatter."""
+
+    can_read_from_local_file: bool = False
+    """Declare whether `read_from_file` is available to this formatter."""
```
Are these `can_read_` attributes intentionally not `ClassVar`?

timj replied:
No. They are class properties.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        self._file_descriptor = file_descriptor
+
+        if ref.isComponent():
+            # Raise or convert to composite ref?
```
I think I lean towards raising, since that's what the docs imply to me and it seems like a bug in higher-level code if a component gets here and it might be confusing to that code if the component is silently ignored.

timj replied:
Turns out that for composite disassembly we pass in the component so it can't raise.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+    def file_descriptor(self) -> FileDescriptor:
+        """File descriptor associated with this formatter (`FileDescriptor`).
+
+        Read-only property.
```
I don't think you need to call this out here, especially since all the other properties are read-only as well.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1058 @@
+            The dataset ref to use when looking in the cache.
+            For single-file dataset this will be the dataset ref directly.
+            If this is disassembled we need the component and the component
+             will be in the `FileDescriptor`.
```
```suggestion
            will be in the `FileDescriptor`.
```

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        component : `str`, optional
+            Component to read from the file. Only used if the `StorageClass`
+            for reading differed from the `StorageClass` used to write the
+            file.
```
Where is the "StorageClass for reading" specified now?  The `DatasetRef` attached to the formatter is now the parent ref, I think.

If the answer is that it's not specified, I think that's great - it'd be a great improvement to just obey `component` and let the "read vs write storage class" stuff focus on storage class conversions.

That does beg the question of what it means to get a component _while_ doing a storage class conversion, but I hope we can get away with banning it, because IIRC QG generation can't handle it, either.

timj replied:
The read vs write storage class definitions haven't moved. They are still in `FileDescriptor`.  The write storage class is the storage class that was used to write the dataset -- it could be the component storage class or the composite depending on whether it was disassembled. You can tell if it was disassembled because `FileDescriptor.component` will be defined.

The read storage class used to only differ from the write one for disassembly but now with storage class conversion they can be different intentionally. I need to make sure that Eli is getting the different value so that if someone stored pandas data frame they can get back a different thing without first reading it as pandas.

timj replied:
I added some extra documentation on the storage class overrides and how they appear to the formatter. I've verified that if you ask for an override it will always turn up in the readStorageClass.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1058 @@
+            for reading differed from the `StorageClass` used to write the
+            file.
+        expected_size : `int`, optional
+            If known the expected size of the resource to read. This can be
```
```suggestion
            If known, the expected size of the resource to read. This can be
```

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        * Derived component requested, read whole file or read relevant
+          component and derive.
+
+        Disassembled Composite:
```
Time to ask this again, I think: should we just drop support for disassembled composites?  The too-many-small-files problem that's led to using Zip archives is another argument against it (unless you see us disassembling and then zipping up the per-component files).

timj replied:
That's too big a change for this ticket and something that should have a broader discussion first.

Disassembling during processing and then zipping when ingesting is an interesting idea and probably would work. Would be a bit slower for the "read the whole thing" case but would be faster for reading a small component (at least until the Exposure replacement arrives).

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+                            in_memory_dataset = self.read_from_stream(zip_fd, component)
+                        except Exception:
+                            zip_fd.close()
+                            raise
```
Could this be written as:
```
with zf.open(path_in_zip) as zip_fd:
    ...
```
If not, maybe do:
```
with contextlib.closing(zf.open(path_in_zip)) as zip_fd:
    ...
```

timj replied:
The first option doesn't work. I read the code and it doesn't do anything with context lib. I will do the second option you suggested.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+                    # URI was remote and file was downloaded
+                    cache_msg = ""
+
+                    if cache_manager.should_be_cached(cache_ref):
```
If we wanted to support caching from one POSIX location to another faster POSIX location (e.g. NFS mount to local scratch), it would be nice if we could do that by just changing what this method returns - I think Princeton might have some interest in that, since we have some very slow POSIX storage here (though a `ChainedDatastore` or even two repos might be better for us).

Right now, I think the "slow POSIX" location would still be interpreted as "local", which would cause two problems:
 - We'd do the read from that slow location first, and then do another slow copy later to move it into the cache.
 - `uri` would be the same as `local_uri`, so we'd never even get to calling `cache_manager.should_be_cached`.

Having written all that, I suppose the other possibility is to look into changing what `ResourcePath.as_local` does for those files; maybe via a `NetworkPosixResourcePath` implementation or something.

And to be clear, I don't think we need that now, and certainly not on this ticket.  But it'd be nice to have a sense for how we might add it.

timj replied:
Isn't it notoriously hard to find out what type of local file system you have? I recall in the past having frustrations trying to work out in a generic way if a directory is on NFS or not.

This has been requested previously in [DM-36150](https://rubinobs.atlassian.net/issues/DM-36150).

TallJimbo replied:
Yeah, I think that's a good reason to not try to fix this via `as_local`; sounds like it'd require inventing a new `file`-like URL schema, which doesn't sound great.  Anyway, I'm happy to leave this for the other ticket.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+                    uri,
+                    msg,
+                    self.name(),
+                )
```
Is this logging redundant with the the timer logging below?

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        ``can_read_from_uri`` is set to `True`.
+
+        It is possible that a cached local file will be given to this method
+        even if it was originally a remote URI. This can happen if the write
```
"if the read", you mean?  And what precisely is "it" in this sentence?  If the formatter can't see what the original URI was at all anyway, it may be better not to mention this.

timj replied:
No. I mean the write. The file can be written to the cache during `butler.put`. The "it" is the URI to the thing being read. It might not be remote even if originally it was a remote URI.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1058 @@
+        Parameters
+        ----------
+        stream : `lsst.resources.ResourceHandleProtocol` or \
+                `typing.BinariyIO`
```
```suggestion
                `typing.BinaryIO`
```

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+
+        Returns
+        -------
+        in_memory_dataset : `typing.Any`
```
I like `object` better than `typing.Any` for at least the docs (as some of the methods already do).  I think it'd also be better for the type annotations unless it leads to a lot of `type: ignore` or casting.

timj replied:
Changing to `object` for the type annotation doesn't seem to make any difference to mypy so I can do that.

timj replied:
Although we use `Any` for `in_memory_dataset` all over the code so are you saying we have to change all those to `object` as well?

timj replied:
Did not do this because it was made more complicated once we returned `NotImplemented`.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        """
+        raise FormatterNotImplementedError("This formatter does not know how to read from a stream.")
+
+    def read_from_local_file(self, local_uri: ResourcePath, component: str | None = None) -> Any:
```
Since this has to be a local file, it might be better to just pass `local_uri.ospath` (i.e. a `str`) as the first argument.

timj replied:
I tried that and then realized that for convenience in some formatters I do `local_uri.read()` and replacing them all with `with open` wrappers will be a bit annoying. I can change them if you prefer it.

timj replied:
I made this change so it uses a `str` for the local read and `ResourcePath` for the `read_from_uri`.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+                    f"Failed to serialize dataset {self.dataset_ref} of "
+                    f"type {get_full_type_name(in_memory_dataset)} to bytes."
+                )
+                raise
```
It might be better to re-raise with `from e` in order to standardize the exception type here so high level code doesn't have to worry about what I/O errors might manifest as.  I don't have a super strong opinion about that, but it's worth thinking about.

timj replied:
Datastore is doing that catch and re-raise.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -75,13 +1109,22 @@ class Formatter(metaclass=ABCMeta):
         Detailed write Recipes indexed by recipe name.
+    **kwargs
+        Additional parameters that can allow parameters
+        from `FormatterV2` to be provided.
```
Does this mean that any concrete V1 formatters now need to accept `**kwargs` as well, because V2 args are going to be passed in?

timj replied:
Yes. But no V1 formatters should have been needing to override the `__init__` so shouldn't notice.

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        Detailed write Recipes indexed by recipe name.
+    **kwargs
+        Additional arguments that will be ignored but allow for
+        `Formatter` V1 parameters to be given.
```
I'm a bit confused as to why we need this.  Under what conditions would we construct a V2 formatter while thinking it's really a V1 formatter?  Seems like we'd just make a `FormatterV1inV2` wrapper instance if we see a V1 formatter configured.

timj replied:
This is because FormatterFactory is creating these instances and it's given `**kwargs` that work for V1 and V2 without it knowing which params are relevant (and that's because the MappingFactory creates the instance and can't know the distinction).

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -797,5 +1884,109 @@ def registerFormatter(
+        recipes: Mapping[str, Any] | None,
+    ) -> Mapping[str, Any] | None:
+        # This should be a class method but a class method can not work
+        # for a dynamic shim. Luckily the shim is only used as an instance.
```
Is it possible to make this an instance method across the board, then?

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -797,5 +1884,109 @@ def registerFormatter(
+            return self._formatter.fromBytes(serialized_dataset, component=component)
+
+        location = Location(*local_uri.split())
+        with self._formatter._updateLocation(location):
```
This gets me wondering whether the V2inV1 wrapper would be simpler if it deferred constructing the V1 formatter until its methods are called, and maybe didn't even hold it as an instance attribute.

timj replied:
That `updateLocation` call has always been a hack that we needed for V1 formatter. The way the formatter instance is constructed (from the formatter factory) makes it hard to defer construction.

TallJimbo commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -786,7 +789,9 @@ def _prepare_for_direct_get(
-    def _prepare_for_put(self, inMemoryDataset: Any, ref: DatasetRef) -> tuple[Location, Formatter]:
+    def _prepare_for_put(
+        self, inMemoryDataset: Any, ref: DatasetRef
+    ) -> tuple[Location, Formatter | FormatterV2]:
```
Could you do the wrapping of `Formatter` into `FormatterV1inV2` earlier (i.e. in this mehtod), to simplify downstream code?

TallJimbo commented on python/lsst/daf/butler/_formatter.py:
```diff
@@ -51,12 +63,1034 @@
+        caller will never call `Formatter.read` or `Formatter.write`.
+    ref : `DatasetRef`
+        The dataset associated with this formatter. Should not be a component
+        dataset ref.
```
Which storage class should this have?  A storage class appears at least here and (twice) in the `file_descriptor`, and there are many different potential storage classes and scenarios in play (read/write, component/full-dataset/[dis]assembly, in-formatter coercion, datastore-level coercion).

More than anything I'd like to drop support for any of those scenarios we don't need, but given that we've accumulated a lot historically and probably can't drop anything other than edge-case combinations without a deprecation period, we probably need to document all possible scenarios (of course, with the rules for those as simple as we can make them).

timj replied:
As written it will have the original storage class that the user passed in if the formatter says it can accept that, else it will have the registry storage class. Before I implemented `can_accept` it was always going to be the registry storage class. Having this available could make the file descriptor `readStorageClass` obsolete. Will need to think.

TallJimbo commented on python/lsst/daf/butler/formatters/pickle.py:
```diff
@@ -34,10 +34,35 @@
+        except pickle.PicklingError:
+            data = None
+
+        return data
```
For a simple formatter like this one, where all we need is bytes, I was expecting it'd be better practice to implement `read_from_stream` rather than `read_from_uri`, because I think of the former as more general.  I can see why others might think the opposite, though, since we could in theory add a `ResourcePath` implementation for anything, even though there are stream types for which there is at present no `ResourcePath` implementation (e.g. `io.BytesIO`).  It's not a big deal because I think either is fine, but it might be good to think about what we recommend when there doesn't seem to be an obvious reason to prefer one or the other.

timj replied:
I haven't tried to see if that works specifically for pickles. It definitely doesn't work for many other cases because `readline` is not implemented by `ResourcePath`.

timj replied:
I checked and `pickle.load()` requires a `readline` implementation. I'm starting to think we are going to have to implement that in ResourcePath.

timj replied:
`yaml.load` does seem to work with ResourcePath though it 's much worse performance because when you hand a stream to `yaml.load` it does reads in chunks of 4096 bytes. There is a 8MB yaml file in summit_utils. If I read that up front and then parse it it completes in 3 seconds but if I pass the handle to the loader it takes 24 seconds so reading up front is definitely better.

TallJimbo commented on python/lsst/daf/butler/formatters/typeless.py:
```diff
@@ -0,0 +1,223 @@
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Support for reading and writing files to a POSIX file system."""
```
Module docstring doesn't seem accurate.

TallJimbo commented on python/lsst/daf/butler/formatters/typeless.py:
```diff
@@ -0,0 +1,223 @@
+        log.warning("About to call super read")
+        data = super().read(component, expected_size, cache_manager)
+
+        log.warning("Read data as type *****************************************: %s", type(data))
```
Debug relic.

TallJimbo commented on python/lsst/daf/butler/formatters/typeless.py:
```diff
@@ -0,0 +1,223 @@
+        # component coercing it to its appropriate pytype.
+        data = self._assemble_dataset(data, component)
+
+        log.warning("Got after assembly: %s", type(data))
```
Debug relic.

TallJimbo commented on python/lsst/daf/butler/formatters/typeless.py:
```diff
@@ -0,0 +1,223 @@
+        if data is None and component is None:
+            raise ValueError(f"Unable to read data with URI {self.file_descriptor.location.uri}")
+
+        log.warning("&************************** Got something of type %s", get_full_type_name(data))
```
Debug relic.

TallJimbo commented on python/lsst/daf/butler/formatters/typeless.py:
```diff
@@ -0,0 +1,223 @@
+    This class provides a ``read()`` method that will run `FormatterV2.read`
+    and coerce the return type using a variety of techniques. Use the
+    standard `FormatterV2` methods for reading bytes/files and writing
+    bytes/files.
```
This gets me thinking that we should have different JSON/YAML formatters for different categories of target types (one for things that are natively pydantic models, another for those that have `to_simple` and `from_simple`, etc.).

May be too late for `PropertySet` vs. `TaskMetadata`, of course, but if we'd had different formatters earlier that transition would have been easier to handle, and I'd like to do better in future similar transitions.

timj replied:
We could have a PydanticFormatter implementation explicitly. This wouldn't have solved the PropertySet vs TaskMetadata problem though because those are entirely different types and the fix there would have been to have a specialist class from the beginning for task metadata that might have internally used a PropertySet rather than using a generic container class at the start.

TallJimbo commented on doc/lsst.daf.butler/formatters.rst:
```diff
@@ -228,69 +228,69 @@ Formatters
-They must though declare all the file extensions that they can support.
-This allows the datastore to ensure that the image being ingested has not obviously been associated with a formatter that does not recognize it.
+``read_from_local_file``
+    The ``read_from_local_file`` method guarantees to pass in a local file resource.
```
```suggestion
    The ``read_from_local_file`` method is guaranteed to be to passed a local file resource.
```

TallJimbo reviewed (approved):
Looks great.  I've left a lot of comments because there's a lot here, but overall I'm very happy with all of this.

I think my biggest concern is that I'd like to make it really clear what all the storage classes are in the various places they appear in a formatter, for the various scenarios in which a formatter is used, and I could imagine trying to document/test that more rigorously leading to some changes to make it more consistent.  There's a line comment thread on this we can use for further discussion.

timj committed:
Changes from review

timj committed:
Add missing docstring

timj committed:
Return NotImplmented rather than raising in formatter

This is more akin to how dunder methods work in general and
is a reasonable approach when we have three options and want to
try each.

timj committed:
Document exceptions raised by formatter V2

timj committed:
Now defer calling read_from_uri if whole file is to be read and cached

timj committed:
Documentation clean up

timj commented on python/lsst/daf/butler/formatters/json.py:
```diff
@@ -34,90 +34,45 @@
-        """
         try:
-            data = json.loads(serializedDataset)
+            data = json.loads(json_bytes)
```
The discussion on https://docs.pydantic.dev/latest/concepts/json/#json-parsing implies that 

1. Trying `pytype.model_validate_json()` here might be significantly faster.
2. We might want to consider using `jiter` for any other JSON if it's meant to be so much faster.
3. @TallJimbo might be right that a special PydanticJsonFormatter is better (although we still have to let JsonFormatter support Pydantic because of backwards compatibility).

@ktlim I assume that `jiter` is part of rubinenv 9 because of the pydantic dependency but maybe it should be explicit rather than explicit if I want to go with point 2 above.

timj replied:
Oh. The new json parser is part of pydantic itself: https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core.from_json

timj replied:
Comparison reading 100MB JSON file:
```
In [8]: %timeit json.loads(data)
539 ms ± 6.83 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: %timeit from_json(data)
266 ms ± 5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
So factor of two improvement but I don't know how many JSON files we are going to be writing that are around that size or larger. (YAML reading the same data is a disaster @ nearly 2 minutes).

timj committed:
Use the pydantic JSON parser for speed

timj committed:
Use lists for __all__ rather than string

timj committed:
Change read_from_local_file to use a path string and not URI

timj committed:
Refresh pre-commit

timj committed:
Change order for is_dataclass check to aid mypy 1.11

This allows the type ignore to be removed.

timj committed:
Remove logging that was duplicated

timj committed:
Disable default acceptance on put

ci_middleware has mocks that look like they are all compatible
with the requested python type when they aren't so by default
a formatter can't naively compare types and decide that since
they match no conversion is required. Always return False for
Formatter.can_accept and force type coercion (which works
with the mocks because of subclassing). Formatter can't know
about mocks (which are in pipe_base) and returning False
doesn't change anything in the default case because conversion
will be a no-op.

The mocks also track file conversions so if conversions are
bypassed the tests confirming conversions won't pass.

timj committed:
Clarify in docs that read_from_local_file uses a path not URI

timj committed:
Fix some errors in docstring markup

timj committed:
Allow for a storage class converter to be a builtin type

timj committed:
Remove now unused cache_ref parameter

This is now all handled internally in the formatter.

timj committed:
Ensure that component override storage class is forwarded when disassembled

timj committed:
Add some notes on storage class overrides for formatters
